Enhance traversal docs

barakatzir opened the pull request:
<!--
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I ran rustfmt locally
- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->
This PR fixes https://github.com/Qiskit/rustworkx/issues/1361.
It add some elaboration to the traversal BFS and DFS functions, mainly with a more concrete pseudo-code.
I also fixed small doc errors in betweenness function's docstrings.
In the issue I also suggested doing a face-lift to the dijkstra_search. I can add this in to this PR in the style of the other docstrings changes I'm proposing If you'd like.

- [v] I ran the `nox -e docs`  and `cargo doc` to see that the result is good.
- [v] I have read the CONTRIBUTING document.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1375) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1375) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/rustworkx?pullRequest=1375) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1375) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1375) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/rustworkx?pullRequest=1375) it.</sub>

IvanIsCoding commented:
Thanks for submitting this! I will try to review it when I can, but again this is a welcome contribution.

IvanIsCoding commented on rustworkx/__init__.py:
```diff
@@ -1491,40 +1491,63 @@ def cartesian_product(
-          end while
-
-    If an exception is raised inside the callback function, the graph traversal
+        def BFSIterator(G, I, F):
```
For Python I think it makes more sense to talk about `BFS` than `BFSIterator`. I'd rewrite it to keep the old signature ` BFS(G, s)`

IvanIsCoding commented on rustworkx/__init__.py:
```diff
@@ -1565,38 +1633,68 @@ def tree_edge(self, edge):
 @_rustworkx_dispatch
 def dfs_search(graph, source, visitor):
-    """Depth-first traversal of a directed/undirected graph.
+    """Iterative depth-first traversal of a directed/undirected graph.
```
We could have a recursive DFS with something like https://crates.io/crates/recursive these days, so I am not opposed to saying this.

With that in mind, DFS order is implementation dependent and even the order we iterate edges is not promised to be stable so it is what it is.

barakatzir replied:
I copied the `Iterative` prefix from the `rustworkx-core` since it made sense to me to differentiate this method from the DFS algorithm which has a single source vertex.
I didn't know about DFS (the CLRS ref you sent me was helpful!) but my friends who do know (both the algo and of CLRS) were confused a bit by how several source vertices are incorporated.
If the implementation detail of how several source vertices are incorporated into DFS is not guaranteed, the docs should probably mention that as well.

IvanIsCoding replied:
I will double check Georgios’ implementation in the source code and check.

I myself forgot the specifics, either the user passes it as an input or we pick an arbitrary node I think

IvanIsCoding replied:
I think if you rewrite the pseudo code to be similar to the implementation, users will like it. Go for it

barakatzir replied:
I'll try to get to it tomorrow.

IvanIsCoding commented on rustworkx/__init__.py:
```diff
@@ -1565,38 +1633,68 @@ def tree_edge(self, edge):
-          color[u] := BLACK                     finish vertex u
-
-    If an exception is raised inside the callback function, the graph traversal
+        def DFSIterator(G, I, F):
```
Same comment from BFS

IvanIsCoding commented on rustworkx/__init__.py:
```diff
@@ -1565,38 +1633,68 @@ def tree_edge(self, edge):
+          color[s] = GRAY
+          time += 1
+          F.Discover(s, time)
+          Q = [s]                             # LIFO vertex queue
```
You can call it a stack

IvanIsCoding reviewed (approved):
This looks good to me with the caveat that I think for Python docstrings we should focus on the methods that highlight `BFS`/`DFS` only. I'd drop the iterator details

barakatzir commented:
I reverted to the pseudo-code styling of the algorithms. I couldn't simplify the algorithms while staying similar to the implementation, so I thought going with pseudo-code conveys better that this does not follow the implementation.

I also redid the Dijkstra search as it was similarly unclear (to me) and added an example.

I fixed a minor type annotation issue in the universal `dijkstra_search`.

IvanIsCoding reviewed (approved):
Thanks for the documentation improvements!

I will merge #1380 with this and then lets check the docs. I revised but there are multiple repetitions of the same docstring (unfortunately, that is how the docs work). So let's see the end result in the website and judge. I have a feeling it will be better
